jstz_api/src/encoding/text_decoder.rs

```diff
@@ -0,0 +1,381 @@
+            obj.get(js_string!("ignoreBOM"), context)?
+                .try_js_into(context)?
+        } else {
+            def.ignore_bom
+        };
+        Ok(Self { fatal, ignore_bom })
+    }
+}
+
+#[derive(Trace, Finalize, Default)]
+pub struct TextDecodeOptions {
+    stream: bool,
+}
+
+impl TryFromJs for TextDecodeOptions {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        let obj = value.as_object().ok_or_else(|| {
+            JsError::from_native(JsNativeError::typ().with_message("Expected `JsObject`"))
+        })?;
+        let def = TextDecodeOptions::default();
+        let stream = if obj.has_property(js_string!("stream"), context)? {
+            obj.get(js_string!("stream"), context)?
+                .try_js_into(context)?
+        } else {
+            def.stream
+        };
+        Ok(Self { stream })
+    }
+}
+
+impl TextDecoder {
+    fn try_from_js<'a>(value: &'a JsValue) -> JsResult<GcRefMut<'a, Object, Self>> {
+        value
+            .as_object()
+            .and_then(|obj| obj.downcast_mut::<Self>())
+            .ok_or_else(|| {
+                JsNativeError::typ()
+                    .with_message(
+                        "Failed to convert js value into rust type `TextDecoder`",
+                    )
+                    .into()
+            })
+    }
+
+    //  https://encoding.spec.whatwg.org/#dom-textdecoder
+    fn constructor(
+        label: Option<String>,
+        options: Option<TextDecoderOptions>,
+    ) -> Result<TextDecoder, ()> {
+        //  Handle optional parameters
+        let label = label.unwrap_or("utf-8".to_string());
+        let options = options.unwrap_or(TextDecoderOptions::default());
+        //  The new TextDecoder(label, options) constructor steps are:
+        //  1. Let encoding be the result of getting an encoding from label.
+        //  2. If encoding is failure or replacement, then throw a RangeError.
+        let encoding =
+            encoding_rs::Encoding::for_label_no_replacement(label.as_bytes()).ok_or(())?;
+        Ok(TextDecoder {
+            //  3. Set this's encoding to encoding.
+            encoding: encoding,
+            //  4. If options["fatal"] is true, then set this's error mode to "fatal".
+            error_mode: if options.fatal {
+                "fatal".to_string()
+            } else {
+                "replacement".to_string()
+            },
+            decoder: if options.ignore_bom {
+                encoding.new_decoder_without_bom_handling()
+            } else {
+                encoding.new_decoder()
+            },
+            io_queue: Vec::new(),
+            //  5. Set this's ignore BOM to options["ignoreBOM"].
+            ignore_bom: options.ignore_bom,
+            //  A TextDecoder object has an associated do not flush, which is a boolean, initially false.
+            do_not_flush: false,
+        })
+    }
+    fn encoding(context: &mut Context<'_>) -> Accessor {
```

Comment:
Style: Line breaks 😉

Reply:
Also generally we'd split this between `TextEncoder` and `TextEncoderClass`. e.g. 
```rust
fn encoding(&self) -> String {
  self.encoding().name().to_lowercase()
}
```

And then define the accessor in `TextEncoderClass`

Reply:
The idea is at somepoint we'll be able to do:
```rust
#[boa(method)]
fn encoding(&self) -> String {
  self.encoding().name().to_lowercase()
}
```

and remove `TextEncoderClass` entirely

Reply:
> Style: Line breaks 😉

ugh, where ? I would hope using the formatter is enough :(

Reply:
for the rest, done in both decoder and encoder

Reply:
> ugh, where ? I would hope using the formatter is enough :(

I will investigate, but currently `cargo fmt` doesn't enforce this